app/src/main/java/org/listenbrainz/android/util/JobQueue.kt

```diff
@@ -0,0 +1,169 @@
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.util.LinkedList
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.coroutines.CoroutineContext
+
+class JobQueue(
+    private val dispatcher: CoroutineDispatcher,
+    private val log: Log = Log.Companion
+) {
+    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
+        log.e("JobQueue Exception: ${e.message}")
+    }
+    private val scope = CoroutineScope(SupervisorJob() + dispatcher)
+    
+    /** `Token: List<QueueJob>` map. To support jobs with same token, we are creating list for
+    * a specified token. */
+    private val jobsMap = HashMap<Any?, LinkedList<QueueJob>>()
+    private val queue = Channel<QueueJob>(Channel.UNLIMITED)
+    
+    // Locks.
+    private val syncLock = Mutex()
+    private val mapLock = Mutex()
+    
+    init {
+        with(scope) {
+            // Execution queue
+            launch(exceptionHandler) {
+                queue.receiveAsFlow().collect { queueJob ->
+                    ensureActive()
+                    
+                    if (!queueJob.cancelled.get()) {
+                        queueJob.job.join()
+                        log.d("Job with token: ${queueJob.token} completed.")
+                    } else {
+                        queueJob.job.cancelAndJoin()
+                        log.d("Job with token: ${queueJob.token} was cancelled.")
+                    }
+                    
+                    // Remove the job from jobMap
+                    mapLock.withLock {
+                        jobsMap[queueJob.token]?.let { jobList ->
+                            // Remove reference to the QueueJob
+                            jobList.remove(queueJob)
+                            // Remove token from map if empty.
+                            if (jobList.isEmpty()) {
+                                jobsMap.remove(queueJob.token)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    fun post(
+        token: Any? = null,
+        context: CoroutineContext = dispatcher,
+        block: suspend CoroutineScope.() -> Unit
+    ) {
+        scope.launch {
+            val queueJob = createQueueJob(token, context, block)
+            addJobToMap(queueJob)
+            sendJobToQueue(queueJob)
+        }
+    }
+    
+    fun postDelayed(
+        delayMillis: Long,
+        token: Any? = null,
+        context: CoroutineContext = dispatcher,
+        block: suspend CoroutineScope.() -> Unit
+    ) {
+        scope.launch {
+            val queueJob = createQueueJob(token, context, block)
```

Comment:
This logic here is redundant. Just do delay first and then call post() directly (the method above)

Reply:
My first approach was this only, but we need to add the job to `jobsMap` before delay so that we can mark the job as cancelled even when the delay is taking place.

Reply:
Cool, makes sense